### DISCUSSION TITLE
Is it safe to share the cache folder over NFS?

### DISCUSSION THREAD

**Discussion by t0ny-peng** [ccache/ccache#795] [Questions & answers] [answered]
I noticed that NFS is mentioned in the manual. However during the test we found that the cache hit rate varies a lot randomly among different machines who shared an NFS as the cache folder, even if they are building the same code with the same everything else. The temp folder, however, is placed on each machine, not across NFS. 

I started to wonder if the problem is due to some kind of data race. Has anybody seen similar symptom working with ccache + NFS?

Thanks.

**Comment by jrosdahl** [comment, marked as answer]
Which version of ccache do you use?

Regarding safety: Ccache in general only depends on file system semantics like atomic renaming and by design doesn't need locks. The exception is the writing of statistics files which uses locks by atomically creating symlinks. If NFS it correctly set up it should to my knowledge be safe.

Regarding cache misses: This is a question that is hard to answer without access to the actual build environment. Fortunately there is (since ccache 3.5) a good way for you to troubleshoot unexpected cache misses yourself: https://ccache.dev/manual/4.2.html#_cache_debugging

> even if they are building the same code with the same everything else.

Just a control question: Does "everything else" also mean exactly the same compiler, including modification timestamps (mtimes), etc?
